specs/Picker-API_spec.md

```diff
@@ -0,0 +1,115 @@
+# Background
+
+The current FilePicker API for UWP has several limitations. Some of the top limitations 
+are as described below:
+
+**1. Unable to pick files and folders in a single instance of the Picker dialog**
+
+   The Picker dialog allows developers to specify picker mode to be single file or multiple
+   files. It does not however, allow developers to specify a mode to pick both files and folders.
+   There has been feedback indicating that this behaviour is limiting, requiring developers
+   to build their own picker experience. 
+
+**2. Unable to pick multiple folders.**
+
+   The Picker dialog (FolderPicker) does not allow users to pick multiple folders. The developer
+   experience currently is to select a single folder and repeat the dialog for more folders. This 
+   is a less than ideal experience for application developers.
+
+The Reunion picker dialog is going to address the above issues. This Reunion API will be based upon the below
+APIs in the Windows SDK. The Reunion Picker API will remove the deprecated APIs and add additional methods
+for the functionality: File and folder picker, Multiple folder picker.
+
+- Windows.Storage.Pickers.FileOpenPicker (https://docs.microsoft.com/en-us/uwp/api/Windows.Storage.Pickers.FileOpenPicker)
+- Windows.Storage.Pickers.FileSavePicker (https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers.filesavepicker)
+- Windows.Storage.Pickers.FolderPicker (https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers.folderpicker) 
+
+# Description
+
+The goal of this API is to address the gaps in the existing File/Folder picker APIs. The API surface is similar to the existing
+Picker APIs in Windows SDK with additional methods to support File + Folder picker and multiple folder picker. The API removes the
+deprecated methods from Windows SDK.
```

Comment:
When returned to an app, I _assume_ the app has all the same access to these objects as if they'd been picked by the platform API instead?  Future access, available through `CreateFileFromApp`/`FindFirstFileExFromApp`, etc.

Reply:
That's not the case now for the pickers and the behavior should remain the same. E.g. The items returned from the picker are available temporarily and are not available by PATH or any other means unless added to the MRU or FutureAccessList. In particular the implementation must NOT internally add items to either list as the contents of those lists are _owned_ by the application - not the system. Though we may want to take the opportunity here to include the user intent that access is supposed to only be one time to and block adding the item to the MRU/FAL @ptorr-msft may have some thoughts on that as he's talked about that before.

Reply:
> Though we may want to take the opportunity here to include the user intent that access is supposed to only be one time to and block adding the item to the MRU/FAL

This in turn begs the question: If you are blocked from adding a path selected via the picker to the MRU/FAL, then how are you supposed to populate those lists in the first place? (Especially in an AppContainer app, where the only way to get access to arbitrary paths is to use the picker.) This is a bizarre limitation, especially since the platform pickers don’t do this. If we go through with this idea, then apps will *lose* functionality by moving to Reunion. This does not sound like a good idea to me.

Reply:
Since we're blue-skying here... the ultimate goal is for the duration of the consent to be specified by the user -- "Until I close the app" or "Always" or "Only when the app is in the foreground" etc. -- and also the scope of the grant (read-only vs. read/write) to be under their control. The apps should not be required to use FAL in order to maintain access once we do this; it should "just work."

Reply:
I agree. The picker should return a vector if IStorageItem with no guarantees on the underlying object.

---
In reply to: [458438812](https://github.com/microsoft/ProjectReunion/pull/99#discussion_r458438812) [](ancestors = 458438812)